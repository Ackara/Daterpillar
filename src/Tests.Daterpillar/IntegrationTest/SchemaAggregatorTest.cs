using ApprovalTests;
using ApprovalTests.Namers;
using ApprovalTests.Reporters;
using Gigobyte.Daterpillar;
using Gigobyte.Daterpillar.Aggregation;
using Gigobyte.Daterpillar.TextTransformation;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System.Data;
using System.IO;

namespace Tests.Daterpillar.IntegrationTest
{
    [TestClass]
    [UseApprovalSubdirectory(nameof(ApprovalTests))]
    [UseReporter(typeof(DiffReporter), typeof(ClipboardReporter))]
    public class SchemaAggregatorTest
    {
        [TestMethod]
        [Owner(Test.Dev.Ackara)]
        [TestCategory(Test.Trait.Integration)]
        public void FetchSchema_should_build_a_schema_object_from_a_active_mssql_database()
        {
            // Arrange
            var connectionString = Test.ConnectionString.MSSQL;
            IgnoreTestIfDbConnectionIsUnavailable(SampleData.CreateMockSchema(), new System.Data.SqlClient.SqlConnection(connectionString), new MSSQLTemplate());

            var sut = new MSSQLSchemaAggregator(new System.Data.SqlClient.SqlConnection(connectionString));

            // Act
            var remoteSchema = sut.FetchSchema();
            var memoryStream = new MemoryStream();
            remoteSchema.WriteTo(memoryStream);

            var validator = new Utilities.XmlValidator();
            validator.Load(memoryStream);

            // Assert
            Approvals.VerifyBinaryFile(memoryStream.ToArray(), "xml");
            Assert.IsTrue(validator.XmlDocIsValid);
        }

        [TestMethod]
        [Owner(Test.Dev.Ackara)]
        [TestCategory(Test.Trait.Integration)]
        public void FetchSchema_should_build_a_schema_object_from_a_active_mysql_database()
        {
            // Arrange
            var connectionString = Test.ConnectionString.MySQL;
            IgnoreTestIfDbConnectionIsUnavailable(SampleData.CreateMockSchema(), new MySql.Data.MySqlClient.MySqlConnection(connectionString), new MySQLTemplate());

            var sut = new MySQLSchemaAggregator(new MySql.Data.MySqlClient.MySqlConnection(connectionString));

            // Act
            var remoteSchema = sut.FetchSchema();
            var memoryStream = new MemoryStream();
            remoteSchema.WriteTo(memoryStream);

            var validator = new Utilities.XmlValidator();
            validator.Load(memoryStream);

            // Assert
            Approvals.VerifyBinaryFile(memoryStream.ToArray(), "xml");
            Assert.IsTrue(validator.XmlDocIsValid);
        }

        [TestMethod]
        [Owner(Test.Dev.Ackara)]
        [TestCategory(Test.Trait.Integration)]
        public void FetchSchema_should_build_a_schema_object_from_a_active_sqlite_database()
        {
            // Arrange
            var schema = SampleData.CreateMockSchema();
            var connStr = new System.Data.SQLite.SQLiteConnectionStringBuilder(Test.ConnectionString.SQLite);
            if (!File.Exists(connStr.DataSource)) System.Data.SQLite.SQLiteConnection.CreateFile(connStr.DataSource);
            IgnoreTestIfDbConnectionIsUnavailable(schema, new System.Data.SQLite.SQLiteConnection(connStr.ConnectionString), new SQLiteTemplate());

            var sut = new SQLiteSchemaAggregator(new System.Data.SQLite.SQLiteConnection(connStr.ConnectionString));

            // Act
            var remoteSchema = sut.FetchSchema();
            var memoryStream = new MemoryStream();
            remoteSchema.WriteTo(memoryStream);

            var validator = new Utilities.XmlValidator();
            validator.Load(memoryStream);

            // Assert
            Approvals.VerifyBinaryFile(memoryStream.ToArray(), "xml");
            Assert.IsTrue(validator.XmlDocIsValid);
        }

        #region Private Members

        private static void IgnoreTestIfDbConnectionIsUnavailable(Schema schema, IDbConnection connection, ITemplate template)
        {
            bool wasNotSuccessful = !SampleData.TryCreateDatabase(connection, schema, template);
#if DEBUG
            if (wasNotSuccessful) Assert.Inconclusive();
#else
            if (wasNotSuccessful) throw new System.ArgumentException($"Cannot connection to '{connection.Database}' database.");
#endif
        }

        #endregion Private Members
    }
}