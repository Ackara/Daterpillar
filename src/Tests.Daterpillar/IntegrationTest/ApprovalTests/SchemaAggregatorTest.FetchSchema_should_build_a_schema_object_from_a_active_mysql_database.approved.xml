<?xml version="1.0"?>
<schema xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" name="daterpillar" xmlns="http://api.gigobyte.com/schema/v1/xddl.xsd">
  <table name="album">
    <comment />
    <column name="Artist_Id" nullable="false" autoIncrement="false">
      <comment />
      <dataType scale="10" precision="0">int</dataType>
    </column>
    <column name="Name" nullable="false" autoIncrement="false">
      <comment />
      <dataType scale="64" precision="0">varchar</dataType>
    </column>
    <column name="Release_Date" nullable="true" autoIncrement="false">
      <comment />
      <dataType scale="0" precision="0">date</dataType>
    </column>
    <foreignKey name="album_ibfk_1" localColumn="Artist_Id" foreignTable="artist" foreignColumn="Id">
      <onUpdate>CASCADE</onUpdate>
      <onDelete>CASCADE</onDelete>
    </foreignKey>
    <index type="primaryKey" unique="false" name="PRIMARY">
      <columnName order="ASC">Artist_Id</columnName>
      <columnName order="ASC">Name</columnName>
    </index>
  </table>
  <table name="artist">
    <comment />
    <column name="Id" nullable="false" autoIncrement="true">
      <comment />
      <dataType scale="10" precision="0">int</dataType>
    </column>
    <column name="Name" nullable="false" autoIncrement="false">
      <comment />
      <dataType scale="64" precision="0">varchar</dataType>
    </column>
    <column name="Bio" nullable="true" autoIncrement="false">
      <comment />
      <dataType scale="65535" precision="0">text</dataType>
    </column>
    <index type="index" unique="true" name="artist_idx1">
      <columnName order="ASC">Name</columnName>
    </index>
  </table>
  <table name="genre">
    <comment />
    <column name="Id" nullable="false" autoIncrement="true">
      <comment />
      <dataType scale="10" precision="0">int</dataType>
    </column>
    <column name="Name" nullable="false" autoIncrement="false">
      <comment />
      <dataType scale="64" precision="0">varchar</dataType>
    </column>
  </table>
  <table name="song">
    <comment />
    <column name="Id" nullable="false" autoIncrement="true">
      <comment />
      <dataType scale="10" precision="0">int</dataType>
    </column>
    <column name="Album_Id" nullable="false" autoIncrement="false">
      <comment />
      <dataType scale="10" precision="0">int</dataType>
    </column>
    <column name="Artist_Id" nullable="false" autoIncrement="false">
      <comment />
      <dataType scale="10" precision="0">int</dataType>
    </column>
    <column name="Genre_Id" nullable="false" autoIncrement="false">
      <comment />
      <dataType scale="10" precision="0">int</dataType>
    </column>
    <column name="Name" nullable="false" autoIncrement="false">
      <comment />
      <dataType scale="256" precision="0">varchar</dataType>
    </column>
    <column name="Length" nullable="true" autoIncrement="false">
      <comment />
      <dataType scale="4" precision="2">decimal</dataType>
    </column>
    <column name="Price" nullable="false" autoIncrement="false">
      <comment />
      <dataType scale="12" precision="2">decimal</dataType>
    </column>
    <column name="On_Device" nullable="false" autoIncrement="false">
      <comment />
      <dataType scale="3" precision="0">bool</dataType>
      <modifier>0</modifier>
    </column>
    <foreignKey name="song_ibfk_1" localColumn="Genre_Id" foreignTable="genre" foreignColumn="Id">
      <onUpdate>CASCADE</onUpdate>
      <onDelete>CASCADE</onDelete>
    </foreignKey>
    <foreignKey name="song_ibfk_2" localColumn="Artist_Id" foreignTable="artist" foreignColumn="Id">
      <onUpdate>CASCADE</onUpdate>
      <onDelete>CASCADE</onDelete>
    </foreignKey>
    <index type="index" unique="false" name="Artist_Id">
      <columnName order="ASC">Artist_Id</columnName>
    </index>
    <index type="index" unique="false" name="Genre_Id">
      <columnName order="ASC">Genre_Id</columnName>
    </index>
    <index type="index" unique="false" name="song_idx">
      <columnName order="ASC">Name</columnName>
    </index>
  </table>
</schema>